tests/models/test_evaluation.py

```diff
@@ -0,0 +1,188 @@
+    expected_metrics = {
+        "accuracy_score": expected_accuracy_score,
+    }
+    expected_saved_metrics = {
+        "accuracy_score_on_iris_dataset": expected_accuracy_score,
+    }
+
+    expected_artifact = confusion_matrix(y_true, y_pred)
+
+    with mlflow.start_run() as run:
+        eval_result = evaluate(
+            classifier_model,
+            "classifier",
+            iris_dataset,
+            run_id=None,
+            evaluators="dummy_evaluator",
+        )
+
+    artifact_name = "confusion_matrix_on_iris_dataset.csv"
+    saved_artifact_path = get_local_artifact_path(run.info.run_id, artifact_name)
+
+    _, saved_metrics, _, saved_artifacts = get_run_data(run.info.run_id)
+    assert saved_metrics == expected_saved_metrics
+    assert saved_artifacts == [artifact_name]
+
+    assert eval_result.metrics == expected_metrics
+    returned_confusion_matrix_artifact = eval_result.artifacts[artifact_name]
+    assert np.array_equal(returned_confusion_matrix_artifact.content, expected_artifact)
+    assert returned_confusion_matrix_artifact.uri == get_artifact_uri(
+        run.info.run_id, artifact_name
+    )
+    assert np.array_equal(
+        returned_confusion_matrix_artifact.load(saved_artifact_path), expected_artifact
+    )
+
+
+def test_regressor_evaluate(regressor_model, iris_dataset):
+    with mlflow.start_run() as run:
+        mlflow.sklearn.log_model(regressor_model, "reg_model")
+        regressor_model_uri = get_artifact_uri(run.info.run_id, "reg_model")
+
+    y_true = iris_dataset.labels
+    regressor_model = mlflow.pyfunc.load_model(regressor_model_uri)
+    y_pred = regressor_model.predict(iris_dataset.data)
+    expected_mae = mean_absolute_error(y_true, y_pred)
+    expected_mse = mean_squared_error(y_true, y_pred)
+    expected_metrics = {
+        "mean_absolute_error": expected_mae,
+        "mean_squared_error": expected_mse,
+    }
+    expected_saved_metrics = {
+        "mean_absolute_error_on_iris_dataset": expected_mae,
+        "mean_squared_error_on_iris_dataset": expected_mse,
+    }
+
+    for model in [regressor_model, regressor_model_uri]:
+        with mlflow.start_run() as run:
+            eval_result = evaluate(
+                model,
+                "regressor",
+                iris_dataset,
+                run_id=None,
+                evaluators="dummy_evaluator",
+            )
+        _, saved_metrics, _, _ = get_run_data(run.info.run_id)
+        assert saved_metrics == expected_saved_metrics
+        assert eval_result.metrics == expected_metrics
+
+
+def test_dataset_name():
+    X, y = get_iris()
+    d1 = EvaluationDataset(data=X, labels=y, name="a1")
+    assert d1.name == "a1"
+    d2 = EvaluationDataset(data=X, labels=y)
+    d2.name == d2.hash
+
+
+def test_dataset_hash(iris_dataset, iris_pandas_df_dataset):
+    assert iris_dataset.hash == "49a04c127e5441e0f27e63a325b5fa69"
+    assert iris_pandas_df_dataset.hash == "d6770fd5fffe651cb95e965854920df9"
```

Comment:
Have we verified that this works across Python versions? Also, can we add a test for Spark DataFrames?

Reply:
I revisit the hash calculation.
the hash calc over `numpy_array.tobytes` is risky because the tobytes return raw data in memory, which might change between different platform/environment.

I replace it with `pandas.util.hash_pandas_object`, `pandas.util.hash_array` and `struct.pack` (pack uint64 numbers into platform-independent bytes)

Now I can confirm the latest hash calculation is deterministic and platform-independent. if so I think we don't need test for "verified that this works across Python versions"